dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/RateLimiter.java

```diff
@@ -0,0 +1,224 @@
+     * @see RateLimiter#bucket
+     */
+    public void consume() throws InterruptedException {
+        bucket.asBlocking().consume(1);
+    }
+
+    /**
+     * Creating this as inner class as existence of this class outside
+     * RateLimiter class will make a little sense. This class is solely
+     * meant to encapsulate the primitives needed for the RateLimiter
+     * and logic for populating those after tokenizing the input
+     */
+    class RateLimitEvent {
+        private final Pattern eventPattern =
+                Pattern.compile("(^[1-9]\\d{0,18})(-per-)?([1-9]\\d{0,18})*([smh]?)", Pattern.CASE_INSENSITIVE);
+        private final long rateLimit;
+        private final Duration rateLimitDuration;
+        private final Duration tokenAddingRate;
+
+        public RateLimitEvent(long rateLimit, Duration rateLimitDuration, Duration tokenAddingRate) {
+            this.rateLimit = rateLimit;
+            this.rateLimitDuration = rateLimitDuration;
+            this.tokenAddingRate = tokenAddingRate;
+        }
+
+        public RateLimitEvent() {
+            this(0, null, null);
+        }
+
+        public long getRateLimit() {
+            return rateLimit;
+        }
+
+        public Duration getRateLimitDuration() {
+            return rateLimitDuration;
+        }
+
+        public Duration getTokenAddingRate() {
+            return tokenAddingRate;
+        }
+
+        public RateLimitEvent tokenizeAndGetEvent(String input) {
+
+            if (input == null) {
+                throw new UnknownFormatConversionException("Can not tokenize/parse a null object");
+            }
+
+            long rLimit;
+            long duration;
+            Duration rLimitDuration;
+            Duration tokAddingRate;
+            String literalConstant;
+            String rateDuration;
+            String rateDurationUnitChar;
+
+            Matcher matcher = eventPattern.matcher(input);
+
+            if (!matcher.matches()) {
+                throw new UnknownFormatConversionException("input argument is not in valid format and can't be tokenize");
+            } else {
+                /**
+                 * eventPattern regex is divided into 5 groups
+                 * 0 group being the whole expression
+                 * 1 group is numeric rate limit value and will always
+                 *                          be present if exp matched
+                 * 2 is literal constant '-per-' and is optional
+                 * 3 is numeric rate duration and is optional
+                 * 4 is a char(s/m/h) and is optional
+                 */
+                rLimit = Long.parseLong(matcher.group(1)); // will always have a numeric value that fits in Long
+                literalConstant = matcher.group(2); // can be null or empty or '-per-'
+                rateDuration = matcher.group(3);// can be null or empty or numeric value
+                rateDurationUnitChar = matcher.group(4); //can be empty or char s,m,h
```

Comment:
You probably want to test that the expected number of groups exist before calling group.  Otherwise this will throw an IndexOutofBoundsException , https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#group(int)

Reply:
I encountered that when the matcher wouldn't match, so wrapped this code around the condition that we execute this only if the pattern matches. 
But I can still handle the exception anyway. let me know your thoughts

Reply:
So I did a quick test with a regex tester.  See https://www.regexplanet.com/share/index.html?share=yyyypr0qx0r .  Wouldn't an input like `1` cause matcher.matches() to be `true` but having less than the desired number of groups?  Maybe we can add a unit test for this.

Reply:
The number of groups is determined by the regex, not by the matcher. The matcher fills in the data if it founds the match. 
for example 
    Pattern.compile("(^[1-9]\\d{0,18})(-per-)?([1-9]\\d{0,18})*([smh]?)").matcher("rando");
        System.out.println(m.matches());
        System.out.println(m.groupCount());
will print false and 4 despite the pattern doesn't match.  Same for value one. it will match and matcher group 1 will have the required value.

Reply:
Got it.  So in the case of the value "1" group 1 will return 1 but group 2 will return null and so forth.